File: python3/n-ary_tree_preorder_traversal.py
from .model import NaryNode

# Space : O(n)
# Time  : O(n)


class Solution:
    def preorder(self, root: NaryNode) -> List[int]:
        def recursive(root):
            res = []

            if root:
                res.append(root.val)
                if root.children:
                    for node in root.children:
                        res += recursive(node)

            return res

        return recursive(root)

File: golang/n-ary_tree_preorder_traversal.go
package main

import (
	"golang/model"
)

// Space : O(n)
// Time  : O(n)

func getPreorder(root *model.NaryNode) []int {
	res := []int{}

	if root != nil {
		res = append(res, root.Val)
		if len(root.Children) > 0 {
			for _, node := range root.Children {
				res = append(res, getPreorder(node)...)
			}
		}
	}

	return res
}

func preorder(root *model.NaryNode) []int {
	return getPreorder(root)
}
